sled-agent/src/bootstrap/trust_quorum/server.rs

```diff
@@ -0,0 +1,128 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::io;
+use std::net::{Ipv6Addr, SocketAddr, SocketAddrV6};
+
+use slog::Logger;
+use tokio::net::{TcpListener, TcpStream};
+use tokio::task::JoinHandle;
+use vsss_rs::Share;
+
+use super::msgs::Response;
+use crate::bootstrap::{agent::BootstrapError, spdm};
+
+// TODO: Get port from config
+// TODO: Get IpAddr from local router:
+//   See https://github.com/oxidecomputer/omicron/issues/443
+pub const PORT: u16 = 12347;
+
+/// A TCP server over which a secure SPDM channel will be established and an
+/// application level trust protocol will run.
+pub struct Server {
+    log: Logger,
+    share: Share,
+    listener: TcpListener,
+}
+
+impl Server {
+    pub fn new(log: &Logger, share: Share) -> io::Result<Self> {
+        let addr = SocketAddrV6::new(Ipv6Addr::UNSPECIFIED, PORT, 0, 0);
+        let sock = socket2::Socket::new(
+            socket2::Domain::IPV6,
+            socket2::Type::STREAM,
+            Some(socket2::Protocol::TCP),
+        )?;
+        sock.set_only_v6(true)?;
+
+        // Allow rebinding during TIME_WAIT
+        sock.set_reuse_address(true)?;
+
+        sock.bind(&addr.into())?;
+        sock.listen(5)?;
+        sock.set_nonblocking(true)?;
+
+        Ok(Server {
+            log: log.clone(),
+            share,
+            listener: TcpListener::from_std(sock.into())?,
+        })
+    }
+
+    pub async fn run(&mut self) -> Result<(), BootstrapError> {
+        loop {
+            // TODO: Track the returned handles in a FuturesUnordered and log any errors?
+            // Alternatively, maintain some shared state across all
+            // responders that is accessable to the Server.
+            let _ = self.accept().await?;
```

Comment:
1. The `recv` wraps a timeout, which has a unit test for exactly this scenario.
2. Attestation is "internal" to the SPDM protocol. Measurements are signed and exchanged in the `CHALLENGE_AUTH` message and/or the `MEASUREMENTS` message. The measurement values that we expect will come from the SP or some configuration, and when measurements are received the requester will access a trait implementation to verify the measurements, similar to how [signing verification](https://github.com/oxidecomputer/spdm/blob/main/src/requester/challenge.rs#L142-L159) utilizes a [trait](https://github.com/oxidecomputer/spdm/blob/main/src/crypto/pki.rs#L25-L53).  The measurement validation code is not yet written.

Measurement occurs before key exchange and the establishment of a secure channel. Rather than just returning the TCP stream here, we will be returning a secure channel similar to TLS 1.3 with encryption and authentication. The important point is that the same pattern holds, and when the entirety of the SPDM protocol is completed here instead of just the negotiation phase, we will actually have a secure channel to send shares over. The user code will look virtually, if not entirely unchanged here. I could have left out this comment in practice because of this.

You can see the SPDM message flow overview on [page 21 of the spec](https://www.dmtf.org/sites/default/files/standards/documents/DSP0274_1.1.1.pdf#%5B%7B%22num%22%3A22%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C683%2C0%5D). So far, in Omicron, we have only utilized the first 3 message exchanges up to `ALGORITHMS`. However the next 3 exchanges, up to `CHALLENGE_AUTH` are already implemented in the spdm library. We do not yet deal with any exchanged measurements though. 

After `FINISH_RSP` is received, the protocol is completely secured and the user will have a secure session with which to send application messages.